frontend-ssr/components/tableOfContents/TableOfContents.tsx

```diff
@@ -0,0 +1,96 @@
+import React from 'react';
+import styles from './TableOfContents.module.scss';
+
+interface State {
+    activeChapterId: string,
+}
+
+interface Props {
+    chapters: Chapters[],
+}
+
+interface Chapters {
+    title: string,
+    id: string,
+}
+
+class TableOfContents extends React.Component<Props, State> {
+  private observer: any;
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      activeChapterId: props.chapters[0].id,
+    };
+  }
+
+  componentDidMount = () => {
+    const changeNav = (entries) => {
+      entries.forEach((entry) => {
+        let entryId = document.getElementById(entry.target.id);
+        if (entryId !== null && entry.intersectionRatio > 0.5) {
+          // remove old active class
+          document.querySelector(`.${styles.active}`).classList.remove(`${styles.active}`);
+          // find matching link & add active class
+          document.querySelector(`[href="#${entry.target.id}"]`).classList.add(`${styles.active}`);
+        }
+      });
+    };
+
+    const options = {
```

Comment:
After you refactor the code from the `changeNav` function, let's add one small improvement. We need to change the URL when we have a match, otherwise if we refresh the page it will lead us to the old chapter.

To do this you're gonna have to use the NextJS router object, specifically it's `.replace` function. - https://nextjs.org/docs/api-reference/next/router

Be careful, here we have a Class Component so you cannot use the hook. You're gonna have to wrap it with `withRouter`, so make sure you read the docs.